core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -376,12 +381,36 @@ class Partition(val topic: String,
   }
 
   /**
+   * Check and maybe increment the low watermark of the partition;
+   * this function can be triggered when any replica's logStartOffset changed
```

Comment:
My concern is that if we store it, but don't make it accurate, then someone may start using it in the future without realizing that it's only accurate at DeleteRecordRequest time. So, it's better to either store it and make it accurate, or not store it and always compute it accurately when needed.